### ISSUE TITLE
Getting output from ansible runner to python and write it on a logfile

### ISSUE THREAD

**Issue by alessandratesta** [ansible/ansible-runner#951]
Hi Guys,
I'm new to python and to ansible.
I was able to run a playbook using ansible runner in python code
but I am not able to write the output of ansible playbook (that I see on the screen) to a log file.
Can you send me some examples?
Thanks

**Comment by Shrews** [comment]
Hello,

This appears to be a user question, not an issue or a feature request. However, since you're new to both Python and Ansible, I question if you really need to use `ansible-runner` at all.  You could use `ansible-playbook` directly which may make things simpler for you.

If you do wish to continue using `ansible-runner`, the `stdout` is saved in the `artifacts` directory of the runner input directory. You can find more information on that [here](https://ansible-runner.readthedocs.io/en/latest/intro/#runner-artifacts-directory-hierarchy) in the docs.

We do have external support resources on IRC (`libra.chat`) in the `#ansible-runner` and `#ansible` channels that you may find useful (see https://docs.ansible.com/ansible/latest/community/communication.html#communication for more info).

**Comment by ppmx** [comment]
Hello folks. I was trying to solve the same thing as @alessandratesta, so I found this issue and was a bit disappointed tbh from @Shrews's answer. From my point of view this is not a user question but actually an issue, because the code still has no way to write the output to a logfile or even to another stream - at all.

The reason for this lies in two files. It starts with the following block in `runner.py`: https://github.com/ansible/ansible-runner/blob/d2afb912a5f7843b3f1799a51e8869e694fb204f/src/ansible_runner/runner.py#L155-L165

This is the only way to set `stdout_handle` as you see, and it points to either a file in the artifacts directory or to `None`, as you can see. When ansible is running via `subprocess` or `pexpect`, this object is used to write the output of the process to `stdout_handler`.

In addition, `stdout_handle` will write the processes output directly into `stdout`, as you can see in: https://github.com/ansible/ansible-runner/blob/d2afb912a5f7843b3f1799a51e8869e694fb204f/src/ansible_runner/utils/__init__.py#L338-L342

So to summarize: it's not about knowing python, but about using `logging` for example, or giving choice to write to an `io.StringIO` for example.

**Comment by ppmx** [comment]
And for all the folks trying to solve that as well, what you can do is using a custom "event handler":

```python
def my_event_handler(event):
    if (dump := event.get("stdout")):
        sys.stdout.write(dump) # <= do your custom logging here

config = ansible_runner.RunnerConfig(xxx)
config.prepare()
config.suppress_ansible_output = True # to avoid ansible_runner's internal stdout dump

r = ansible_runner.Runner(config=config, event_handler=my_event_handler)
r.run()
```

**Comment by zhangguanzhang** [comment]
https://github.com/ansible/ansible-runner/blob/e818a42d4fb5790f392a3218f816929bf8e23de6/src/ansible_runner/config/_base.py#L283
another way
```python
def my_event_handler(event):
    if (dump := event.get("stdout")):
        sys.stdout.write(dump) # <= do your custom logging here

r = ansible_runner.run(playbook='test.yml',
        event_handler=my_event_handler, 
        settings={"suppress_ansible_output": True},
       #quiet=True,
    )
```
